src/main/java/com/artipie/asto/Remaining.java

```diff
@@ -54,9 +69,25 @@ public Remaining(final ByteBuffer buf) {
      */
     public byte[] bytes() {
         final byte[] bytes = new byte[this.buf.remaining()];
-        this.buf.mark();
+        if (this.restore) {
+            this.buf.mark();
+        }
         this.buf.get(bytes);
-        this.buf.reset();
+        if (this.restore) {
+            this.buf.reset();
+        }
         return bytes;
     }
+
+    /**
+     * Buffer read strategy.
+     * @since 0.14
+     */
+    public interface ReadStrategy {
```

Comment:
@g4s8 @g4s8 This is a nice implementation, but now it is leading to nowhere. You should:
- make the a concrete default implementation of `ReadStrategy` for `Remaining`;
- leave a puzzle to the other implementation (which wouldn't restore the position) be made